### PR TITLE
Wi21 135 blynk integration

### PR DESCRIPTION
**Pull request by marek-szydlowski-wizzdev**

### CODE REVIEW

File: Blynk/README.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+An example web dashboard that might be created can look like this:
+![Example dashboard](Blynk_dashboard.png "KaaIoT dashboard")
+
+## What is Blynk
+[Blynk](https://blynk.io/) is a cloud service provider allowing for building and managing connected hardware: device provisioning, sensor data visualization, remote control with mobile and web applications, over-the-air firmware updates and much more.
+
+## Requirements
+
+### Python 3, PIP
```
Comment (damian-kurek-wizzdev) [inline Blynk/README.md:12]: Move this info to the main readme and remove it from cloud-specific readme

File: Blynk/README.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+After you first log in into your new account on Blynk,Console you'll be probably prompted with a quickstart device. Please close the window as we're going to do it manually.
+
+#### **Template**:
+
+Template will be used for creating your new IoT-Starter device.
+
+ 1. On the left side pane search for **"Templates"** option.
+ 2. Click on **"+ New Template"** with the name and description of your choice. Make sure to use hardware **"ESP32"** and connection type **"WiFi"**. After creating the template you can modify it further adding an image, metadata etc.
+ 3. Now you need to create two **"Datastreams"** that are responsible for accepting telemetry data (temperature and humidity) from our device. To do that go to:<br>
```
Comment (damian-kurek-wizzdev) [inline Blynk/README.md:105]: gi to :<br>

File: Blynk/README.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+ **"Datastreams" --> "+ New Datastream" --> "Virtual Pin"<br>**
+ Please make two virtual pin datastreams for temperature and humidity. **Please note** which pin is responsible for each telemetry type (only number matters, so if you're assigning **V0** to temperature, the pin responsible for the temperature telemetry is no. **0**). Make sure that the **"Data Type"** is set to **"Double"**. You can also add **"Units"** to the pins (Celsius for the temperature and percentage for the humidity) - it is optional though. Please also adjust the **min** and **max** value for each pin (0 - 100 for the humidity and for example 10 - 40 for the temperature) as well as check the **"Thousands seperator"** option.
```
Comment (damian-kurek-wizzdev) [inline Blynk/README.md:107]: NO idea

File: Blynk/README.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+ 4. After creating both virtual pins you can set up a web dashboard after clicking **"Web dashboard"** option. (**"Mobile dashboard"** can be only created in **"Blynk IoT"** mobile app.)
+ 5. Save the changes made to the template.
+
+#### **Device**:
+
+Device will be created from already created template, so all you need to do is:
+
+ 1. Click on the left side pane **"Search"** --> **"+ New Device"** --> **"From template"**
+ 2. Select your template and give some meaningful name to your device.
+ 3. Click on the created device and go to **"Device info"** and on the right side look for **"BLYNK_AUTH_TOKEN"**. It should look like this:<br>
```
Comment (damian-kurek-wizzdev) [inline Blynk/README.md:117]: br

File: MicroPython/src/cloud/Blynk_cloud.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+from cloud.cloud_interface import CloudProvider
+
+
+class Blynk_cloud(CloudProvider):
+    def __init__(self) -> None:
+        if not config.cfg.blynk_auth_token:
+            self.configure_data()
+
+        self.auth_token = config.cfg.blynk_auth_token
+        self.temp_pin = config.cfg.blynk_temp_pin
```
Comment (damian-kurek-wizzdev) [inline MicroPython/src/cloud/Blynk_cloud.py:24]: full names

File: MicroPython/src/cloud/Blynk_cloud.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        Connects to WiFi. It is seperate function from utils one, as Blynk doesn't use MQTT Client.
+        :param sync_time: flag if time is synchronized.
+        :return 
+        """
+        logging.debug("Blynk_cloud/wifi_connect()")
+
+        try:
+            wireless_controller = get_wireless_connection_controller_instance()
+            utils.connect_to_wifi(wireless_controller,
+                                  config.cfg.access_points, sync_time)
```
Comment (damian-kurek-wizzdev) [inline MicroPython/src/cloud/Blynk_cloud.py:130]: enter

File: MicroPython/src/cloud/Blynk_cloud.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+            except Exception:
+                logging.error("Error in disconnecting WiFi controller")
+
+        return wireless_controller
+
+    def publish_data(self, data):
+        wireless_controller = Blynk_cloud.wifi_connect(sync_time=False)
+
+        data = self._format_data(data)
+        temperature, humidity = data.get('temperature'), data.get('humidity')
```
Comment (damian-kurek-wizzdev) [inline MicroPython/src/cloud/Blynk_cloud.py:147]: why get without default/ split into 2 lines

File: MicroPython/src/cloud/Things_cloud.py [hunk 8]
```diff
@@ -10,7 +10,7 @@
 from cloud.cloud_interface import CloudProvider
 
 
-class ThingsBoard(CloudProvider):
+class ThingsBoard_cloud(CloudProvider):
```
Comment (damian-kurek-wizzdev) [inline MicroPython/src/cloud/Things_cloud.py:13]: ThingsBoardCloud

File: MicroPython/scripts/common/cloud_providers.py [hunk 9]
```diff
@@ -5,11 +5,13 @@ class Providers:
...
+    BLYNK = "BLYNK"
 
     @classmethod
     def print_providers(cls) -> str:
-        return "{}, {}, {}".format(cls.AWS, cls.KAA, cls.THINGSBOARD)
+        return "{}, {}, {}, {}".format(
+            cls.AWS, cls.KAA, cls.THINGSBOARD, cls.BLYNK)
 
     @classmethod
     def get_providers(cls) -> Tuple[str, str, str]:
```
Comment (patryk-belka-wizzdev) [inline MicroPython/scripts/common/cloud_providers.py:16]: ->Tuple[str, str, str, str]

File: MicroPython/src/cloud/Blynk_cloud.py (hunk 7 above)
Comment (marek-szydlowski-wizzdev) [reply MicroPython/src/cloud/Blynk_cloud.py:147]: Changed

File: MicroPython/src/common/config.py [hunk 10]
```diff
@@ -86,6 +85,12 @@
 DEFAULT_THINGSBOARD_DEVICE_ID = ''
 DEFAULT_THINGSBOARD_ATTRIBUTES_EXISTS = False
 
+# Blunk stuff
```
Comment (patryk-belka-wizzdev) [inline MicroPython/src/common/config.py:88]: Blynk

File: MicroPython/src/cloud/Blynk_cloud.py [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+import utime
+
+from common import config, utils
+from common.utils import ConnectionError
+from communication.wirerless_connection_controller import (
+    WirelessConnectionController, get_wireless_connection_controller_instance)
+from controller.main_controller_event import MainControllerEventType
+from lib.BlynkLib import Blynk
+
+from cloud.cloud_interface import CloudProvider
```
Comment (damian-kurek-wizzdev) [inline MicroPython/src/cloud/Blynk_cloud.py:15]: imports

**Review by patryk-belka-wizzdev — APPROVED** [review]

**Review by damian-kurek-wizzdev — APPROVED** [review]

**Pull request merged**